smarts/core/utils/traffic_history_serivce.py

```diff
@@ -0,0 +1,153 @@
+        self._receive_data_conn.recv()
+
+    def _fetch_history(
+        self, send_data, request_queue, history_file_path, total_timestep_size
+    ):
+        """prepare 1 batch ahead, when received request, immediately return the previously
+        prepared batch and prepares the next batch.
+        """
+        return_batch = {}
+        while True:
+            historyRange = request_queue.get()
+            assert isinstance(historyRange, RequestHistoryRange)
+            send_data.send(return_batch)
+            return_batch = {}
+            with open(history_file_path, "rb") as f:
+                for index, (t, vehicles_state) in enumerate(
+                    ijson.kvitems(f, "", use_float=True)
+                ):
+                    if (
+                        historyRange.start_index <= index
+                        and index <= historyRange.end_index
+                    ):
+                        return_batch[t] = vehicles_state
+        send_data.close()
+
+    @property
+    def all_timesteps(self):
+        return self._all_timesteps
+
+    @property
+    def traffic_history(self):
+        return {**self._current_traffic_history, **self._prev_batch_history}
+
+    def _prepare_next_batch(self):
+        self._request_queue.put(
+            RequestHistoryRange(
+                start_index=self._range_start,
+                end_index=self._range_start + self._batch_size,
+            )
+        )
+        self._range_start += self._batch_size
+
+    def fetch_history_at_timestep(self, timestep):
+        if timestep not in self._all_timesteps:
+            return {}
+        elif timestep in self.traffic_history:
+            return self.traffic_history[timestep]
+
+        # ask child process to prepare the next batch:
+        self._prepare_next_batch()
+        self._prev_batch_history = self._current_traffic_history
+        # receives the previous batch child process prepared
+        self._current_traffic_history = self._receive_data_conn.recv()
+        if timestep in self._current_traffic_history:
+            return self._current_traffic_history[timestep]
+        # no history exists at requested timestamp
+        return {}
+
+    def fetch_agent_missions(self):
+        vehicle_missions = {}
+        with open(self._history_file_path, "rb") as f:
+            for t, vehicles_state in ijson.kvitems(f, "", use_float=True):
+                for vehicle_id in vehicles_state:
+                    if vehicle_id in vehicle_missions:
+                        continue
+                    vehicle_missions[vehicle_id] = scenario.Mission(
+                        start=scenario.Start(
+                            vehicles_state[vehicle_id]["position"][:2],
+                            scenario.Heading(vehicles_state[vehicle_id]["heading"]),
+                        ),
+                        goal=scenario.EndlessGoal(),
+                        start_time=float(t),
+                    )
```

Comment:
Could you elaborate on dumping this to a blob? What are we dumping exactly? Are we dumping the content in traffic_history.json into a blob file and whenever we need to access the data we access the blob file?

I think this might be costly in terms of performance but should be okay in terms of memory since we are getting `t, vehicles_state` for each timestamp and then overwrites t and vehicles_state with data in next timestamp.

Reply:
The performance cost of this is roughly similar to opening the json file using json.load and iterating through all its contents

Reply:
Also if we dump all the contents in json file into a blob file, then when loading the blob file, the size could be very large and we are loading all the contents into memory, this would kinda defeats the purpose of using ijson here?

Reply:
Sorry, to clarify: when we get the first timestep we have to process a potentially massive history file and will have to reprocess the same vehicle multiple times because we want to find the earliest instance of a vehicle. We can skip looping over the json by taking the resulting `vehicle_missions` here and turning it into a binary. Then we can load those missions back in if available since it is faster.

For small files this will not have much effect but it would see a lot of improvement on massive traffic history files.

I do not think we need to do it now but it should be put up as an issue.

Reply:
If I understood your comment correctly, you were saying this method would be called for every vehicle when we reaches its first timestamp. 
Actually this method `discover_missions_of_traffic_histories` is only called once per run of a scenario. 
It is called in `history_vehicles_replacement_for_imitation_learning.py` 
```
agent_missions = scenario.discover_missions_of_traffic_histories()
for agent_id, mission in agent_missions.items():
   ...
```
I think saving the `vehicle_missions` here would only be a somewhat helpful if there are multiple scenarios with same traffic_history_file, but the difference in performance shouldn't be too big

Reply:
@Gamenot

Reply:
OK, there is a bit more to it. I am assuming the converted NGSIM file looks like this:

```
traffic_history_1 = {
    "0.1": {
        "1": {
            "vehicle_id": "1",
            "vehicle_type": "car",
            "position": [1068.124, 959.455, 0],
            "speed": 5.128410474991252,
            "heading": -4.675796326794897,
            "vehicle_length": 4.22,
            "vehicle_width": 1.72,
        },
        "2": {
            "vehicle_id": "2",
            "vehicle_type": "car",
            "position": [1041.407, 956.583, 0],
            "speed": 2.9067860258367832,
            "heading": 1.4512036732051032,
            "vehicle_length": 4.16,
            "vehicle_width": 1.78,
        },
    },
    "0.2": {
        "1": {
            "vehicle_id": "1",
            "vehicle_type": "car",
            ...
        },
        ...
    },
    ...
}
```

So then the iteration is like this:

```            
for t, vehicles_state in ijson.kvitems(f, "", use_float=True):
    for vehicle_id in vehicles_state:
        if vehicle_id in vehicle_missions:
            continue
```

Then that means that the cause of the `if` statement here is because we revisit vehicles that have already been seen before but they are in a different frame (e.g. `vehicle: 1` in frame 1 and 2... and probably something like frames 3->882.) Looking at the NGSIM files I am seeing files as large as 250MB. The cost of reading a json file near that size is going to be terrible when all we need is the first known instance of the vehicle. 

Ideally we would want this to be dynamic but we seem constrained on that point.

Reply:
@Gamenot  yes, we do just need the first known instance of each vehicle in order to construct the mission. But since it is possible that a new vehicle id appears at a later timestamp, we had to iterate through all the timestamps and the vehicles at each timestamps.  

I will create the new issue.

Reply:
Issue created #640